### PR TITLE
Dragging changes, rebased on develop

### PR DESCRIPTION
**Pull request by rachel-fenichel**
Improved block and workspace dragging and clicking.
Gesture.js coordinates events across a full click or drag.
Draggers handle details of dragging a particular object (i.e. flyout, workspace, or block).
Blocks, fields, and workspaces no longer decide what type of gesture is happening.

New classes:
- block dragger
- workspace dragger
- flyout dragger
- gesture
- dragged connection manager

I will want to squash and merge this into develop, then do a rebuild and merge that separately for clarity.

### CODE REVIEW

File: core/gesture.js [hunk 1]
```diff
@@ -0,0 +1,695 @@
...
+    this.startWorkspace_.showContextMenu_(e);
+  }
+
+  e.preventDefault();
+  e.stopPropagation();
+
+  this.dispose();
+};
+
+// Called externally.
```
Comment (picklesrus) [inline core/gesture.js]: Which function(s) is this comment supposed to apply to?

File: core/gesture.js [hunk 2]
```diff
@@ -0,0 +1,695 @@
...
+ */
+Blockly.Gesture.prototype.handleBlockStart = function(e, block) {
+  goog.asserts.assert(!this.hasStarted_,
+     'Tried to call gesture.handleBlockStart, but the gesture had already been ' +
+     'started.');
+  this.setStartBlock(block);
+  this.mostRecentEvent_ = e;
+};
+
+// Field clicks
```
Comment (picklesrus) [inline core/gesture.js]: proper jsdoc/annotations for the next few methods

File: core/gesture.js [hunk 3]
```diff
@@ -0,0 +1,695 @@
...
+};
+
+// Simple getters
+
+/**
+ * Check whether this gesture is a click on a block.  This should only be called
+ * when ending a gesture (mouse up, touch end).
+ * @return {boolean} whether this gesture was a click on a block.
+ */
+Blockly.Gesture.prototype.isBlockClick_ = function() {
```
Comment (picklesrus) [inline core/gesture.js]: @private annotation

File: core/gesture.js [hunk 4]
```diff
@@ -0,0 +1,695 @@
...
+  var hasStartBlock = !!this.startBlock_;
+  return hasStartBlock && !this.hasExceededDragRadius_ && !this.isFieldClick_();
+};
+
+/**
+ * Check whether this gesture is a click on a field.  This should only be called
+ * when ending a gesture (mouse up, touch end).
+ * @return {boolean} whether this gesture was a click on a field.
+ */
+Blockly.Gesture.prototype.isFieldClick_ = function() {
```
Comment (picklesrus) [inline core/gesture.js]: private on this one and a bunch of the next ones too I think

File: core/gesture.js [hunk 5]
```diff
@@ -0,0 +1,695 @@
...
+
+Blockly.Gesture.prototype.isWorkspaceDrag_ = function() {
+  return this.isDraggingWorkspace_;
+};
+
+Blockly.Gesture.prototype.isBlockDrag_ = function() {
+  return this.isDraggingBlock_;
+};
+
+Blockly.Gesture.prototype.isDragging = function() {
```
Comment (picklesrus) [inline core/gesture.js]: comments missing

**Comment by rachel-fenichel** [comment]
Now that you've taken a look at the code without the built files getting in the way, I'll rebuild and then fix up all of your comments.

**Comment by rachel-fenichel** [comment]
Friendly ping!  Travis is happy with this PR, and most of your comments were on comments.  Is this okay to merge?

**Comment by picklesrus** [comment]
Ship it!

**Pull request merged**